README.md


# rprintf

[![Linux Build Status](https://travis-ci.org/renkun-ken/rprintf.png?branch=master)](https://travis-ci.org/renkun-ken/rprintf) 
[![Windows Build status](https://ci.appveyor.com/api/projects/status/github/renkun-ken/rprintf?svg=true)](https://ci.appveyor.com/project/renkun-ken/rprintf)
[![Coverage Status](https://coveralls.io/repos/renkun-ken/rprintf/badge.svg)](https://coveralls.io/r/renkun-ken/rprintf) 
[![CRAN Version](http://www.r-pkg.org/badges/version/rprintf)](http://cran.rstudio.com/web/packages/rprintf)


rprintf is an adaptive builder for formatted strings. Currently it provides a set of tools for building formatted strings under various replacement rules: 

- C-style formatting with `sprintf`
- Number-based formatting
- Variable-based formatting

The primary goal of this package is to make it easier to produce formatted strings in all popular styles.

## Installation

You can install from CRAN with

```r
install.packages("rprintf")
```

or you can install the latest development version from GitHub with

```r
devtools::install_github("rprintf","renkun-ken")
```

## Examples

The following examples demonstrate how `rprintf` functions works. You should be familiar with how `sprintf` works first. See the [documentation](http://stat.ethz.ch/R-manual/R-devel/library/base/html/sprintf.html) for more details.

### C-style formatting


```r
library(rprintf)
rprintf("Hello, %s", "world")
```

```
[1] "Hello, world"
```

```r
rprintf("%s (%d years old)", "Ken", 24)
```

```
[1] "Ken (24 years old)"
```

```r
rprintf("He is %d but has a height of %.1fcm", 18, 190)
```

```
[1] "He is 18 but has a height of 190.0cm"
```

### Number-based formatting


```r
rprintf("Hello, {1}", "world")
```

```
[1] "Hello, world"
```

```r
rprintf("{1} ({2} years old)","Ken",24)
```

```
[1] "Ken (24 years old)"
```

```r
rprintf("He is {1} but has a height of {2:.2f}cm",18,190)
```

```
[1] "He is 18 but has a height of 190.00cm"
```

```r
rprintf("{1}, {2:.1f}, {3:+.2f}, {2}, {1:.0f}",1.56,2.34,3.78)
```

```
[1] "1.56, 2.3, +3.78, 2.34, 2"
```

```r
rprintf("{2},{1}","x","y")
```

```
[1] "y,x"
```

### Variable-based formatting


```r
rprintf("Hello, $name", name="world")
```

```
[1] "Hello, world"
```

```r
rprintf("$name ($age years old)",name="Ken",age=24)
```

```
[1] "Ken (24 years old)"
```

```r
rprintf("He is $age but has a height of $height:.2fcm",age=18,height=190)
```

```
[1] "He is 18 but has a height of 190.00cm"
```

```r
rprintf("$a, $b:.1f, $c:+.2f, $b, $a:.0f",a=1.56,b=2.34,c=3.78)
```

```
[1] "1.56, 2.3, +3.78, 2.34, 2"
```

For each type of formatting, a specialized function is also provided. `rprintv` only handles named variable-based formatting, and `rprintn` only handles number-based formatting.

## License

This package is under [MIT License](http://opensource.org/licenses/MIT).


NEWS
v0.2
===
Import stringi package instead of stringr

v0.1-2 (2014-08-15)
===
Maintenance update

v0.1-1 (2014-07-12)
===
Support list argument


v0.1 (2014-06-06)
===
rprintf, rprintv, and rprintn functions are implemented.


tests/testthat/test-rprintf.R
context("rprintf: general formatting")

test_that("rprintf works as expected", {
  expect_equivalent(rprintf("%s","a"),"a")
  expect_equivalent(rprintf("$name:s",name="a"),"a")
  expect_equivalent(rprintf("{1:s}","a"),"a")
  expect_equivalent(rprintf("$name",name="a"),"a")
  expect_equivalent(rprintf("{1}","a"),"a")

  p <- list(name="Ken",age=25)
  expect_equivalent(rprintf("$name, $age",p),"Ken, 25")
})


tests/testthat/test-rprintn.R
context("rprintn: number-based formatting")

test_that("rprintn works as expected", {
  expect_equivalent(rprintn("{1:s}","a"),"a")
  expect_equivalent(rprintn("{1:d}",10),"10")
  expect_equivalent(rprintn("{1:.2f}",pi),"3.14")
  expect_equivalent(rprintn("{1:+.2f}",pi),"+3.14")
  expect_equivalent(rprintn("{1: .2f}",pi)," 3.14")
  expect_equivalent(rprintn("{2:s},{1}","a","b"),"b,a")
  expect_equivalent(rprintn("{1},{2:s},{3:.2f},{2},{1}","a","b",1.4),"a,b,1.40,b,a")
  expect_equivalent(rprintn("{{1}}","a"),"{1}")
  expect_equivalent(rprintn("{{1:s}}","a"),"{1:s}")

  p <- list(name="Ken",age=25)
  expect_equivalent(rprintn("{1}, {2}",p),"Ken, 25")
})


tests/testthat/test-rprintv.R
context("rprintv: variable-based formatting")

test_that("rprintv works as expected", {
  expect_equivalent(rprintv("$name:s",name="a"),"a")
  expect_equivalent(rprintv("$num:d",num=10),"10")
  expect_equivalent(rprintv("$pi:.2f",pi=pi),"3.14")
  expect_equivalent(rprintv("$pi:+.2f",pi=pi),"+3.14")
  expect_equivalent(rprintv("$pi: .2f",pi=pi)," 3.14")
  expect_equivalent(rprintv("$name1:s,$name",name="a",name1="b"),"b,a")
  expect_equivalent(rprintv("$a,$b:s,$c:.2f,$b,$a",a="a",b="b",c=1.4),"a,b,1.40,b,a")
  expect_equivalent(rprintv("$$name",name="a"),"$name")

  p <- list(name="Ken",age=25)
  expect_equivalent(rprintn("{1} {2}",p),"Ken 25")
})


R/rprintn.R
#' Build a character vector or list with number-based
#' string formatting
#'
#' The \code{rprintn} function applies number-based formatter to
#' transform the given character vector to specific texts with
#' numbers replaced by a given set of values in correct
#' formats.
#'
#' @param x The character vector or list to be transformed
#' @param ... The arguments that specify the set of values to be
#'   placed
#' @export
#' @examples
#' \dontrun{
#'
#' # Format a single-entry character vector with numbering mechanism
#' rprintf("Hello, {1}", "world")
#' rprintf("{1} ({2} years old)","Ken",24)
#' rprintf("He is {1} but has a height of {2:.2f}cm",18,190)
#' rprintf("{1}, {2:.1f}, {3:+.2f}, {2}, {1:.0f}",1.56,2.34,3.78)
#' rprintf("{2},{1}","x","y")
#' }
#'
rprintn <- function(x,...) {
  args <- makelist(...)
  x <- stringr::str_replace_all(x,"%","%%")
  xs <- unlist(stringr::str_extract_all(x,stringr::perl("(?<!\\{)\\{\\d+(:[\\s\\+\\-\\#\\.\\d]*\\w)?\\}(?!\\})")))
  xss <- as.numeric(stringr::str_replace_all(xs,stringr::perl("(?<!\\{)\\{(\\d+)(:[\\s\\+\\-\\#\\.\\d]*\\w)?\\}(?!\\})"),"\\1"))
  pass1 <- stringr::str_replace_all(x,stringr::perl("(?<!\\{)\\{(\\d+):([\\s\\+\\-\\#\\.\\d]*\\w)?\\}(?!\\})"),"%\\2")
  pass2 <- stringr::str_replace_all(pass1,stringr::perl("(?<!\\{)\\{(\\d+)\\}(?!\\})"),"%s")
  pass3 <- do.call(sprintf,c(list(pass2),args[xss]))
  stringr::str_replace_all(pass3,"\\{\\{(.+)\\}\\}","{\\1}")
}


R/rprintv.R
#' Build a character vector or list with variable-based
#' string formatting
#'
#' The \code{rprintv} function applies variable-based formatter to
#' transform the given character vector to specific texts with
#' named variables replaced by a given set of values in correct
#' formats.
#'
#' @param x The character vector or list to be transformed
#' @param ... The arguments that specify the set of values to be
#'   placed
#' @export
#' @examples
#' \dontrun{
#'
#' # Format a single-entry character vector with variable mechanism
#' rprintf("Hello, $name", name="world")
#' rprintf("$name ($age years old)",name="Ken",age=24)
#' rprintf("He is $age but has a height of $height:.2fcm",age=18,height=190)
#' rprintf("$a, $b:.1f, $c:+.2f, $b, $a:.0f",a=1.56,b=2.34,c=3.78)
#' }
#'
rprintv <- function(x,...) {
  args <- makelist(...)
  x <- stringr::str_replace_all(x,"%","%%")
  xs <- unlist(stringr::str_extract_all(x,stringr::perl("(?<!\\$)\\$[\\w\\.]+(:[\\s\\+\\-\\#\\.\\d]*\\w)?")))
  xss <- stringr::str_replace_all(xs,stringr::perl("(?<!\\$)\\$([\\w\\.]+)(:[\\s\\+\\-\\#\\.\\d]*\\w)?"),"\\1")
  pass1 <- stringr::str_replace_all(x,stringr::perl("(?<!\\$)\\$([\\w\\.]+):(?!\\$)(?!:+)([\\s\\+\\-\\#\\.\\d]*\\w)?"),"%\\2")
  pass2 <- stringr::str_replace_all(pass1,stringr::perl("(?<!\\$)\\$([\\w\\.]+)"),"%s")
  pass3 <- do.call(sprintf,c(list(pass2),args[xss]))
  stringr::str_replace_all(pass3,"\\$\\$","$")
}


R/rprintf.R
#' Build a character vector or list with adaptive string formatting
#'
#' The \code{rprintf} function checks the given character vector or
#' list and applies appropriate formatters that transform it from
#' generic patterns to specific texts with variables and indices
#' as placeholders replaced by a given set of values in correct
#' formats.
#'
#' @param x The character vector or list to be transformed
#' @param ... The arguments that specify the set of values to be
#'   placed
#' @export
#' @examples
#' \dontrun{
#'#' # Format a single-entry character vector with sprintf mechanism
#' rprintf("Hello, %s","world")
#' rprintf("%s (%d years old)","Ken",24)
#' rprintf("He is %d but has a height of %.1fcm",18,190)
#'
#' # Format a single-entry character vector with variable mechanism
#' rprintf("Hello, $name", name="world")
#' rprintf("$name ($age years old)",name="Ken",age=24)
#' rprintf("He is $age but has a height of $height:.2fcm",age=18,height=190)
#' rprintf("$a, $b:.1f, $c:+.2f, $b, $a:.0f",a=1.56,b=2.34,c=3.78)
#'
#' # Format a single-entry character vector with numbering mechanism
#' rprintf("Hello, {1}", "world")
#' rprintf("{1} ({2} years old)","Ken",24)
#' rprintf("He is {1} but has a height of {2:.2f}cm",18,190)
#' rprintf("{1}, {2:.1f}, {3:+.2f}, {2}, {1:.0f}",1.56,2.34,3.78)
#' rprintf("{2},{1}","x","y")
#'
#' # This function also works for character vectors and lists.
#' rprintf(c("%s:%d","$name:$age","{1}:{2}"),name="Ken",age=24)
#' rprintf(c(a="%s:%d",b="$name:$age",c="{1}:{2}"),name="Ken",age=24)
#' rprintf(list("%s:%d","$name:$age","{1}:{2}"),name="Ken",age=24)
#' rprintf(list(a="%s:%d",b="$name:$age",c="{1}:{2}"),name="Ken",age=24)
#'
#' # It also works with list argument for named variables.
#' p <- list(name="Ken",age=24)
#' rprintf("name: $name, age: $age",p)
#' rprintf("name: {1}, age: {2}",p)
#'
#' Note that when the list of arguments are given names,
#' the variable names in format string should be modified.
#' rprintf("name: $arg.name, age: $arg.age", arg = p)
#' }
rprintf <- function(x,...) {
  matches <- do.call(cbind,lapply(patterns,
    function(p) grepl(p,x,perl = TRUE)))
  funs.id <- apply(matches,1L,function(row) which(row)[1L])
  funs <- names(patterns)[funs.id]
  result <- Map(rprintf.match,x,funs,list(list(...)))
  if(is.list(x)) result
  else setnames(unlist(result,use.names = FALSE),names(x))
}
